### PR TITLE
Draft a new E-Mail gateway

### LINKED ISSUE #8: Simplify notification and subscription process (also for tags)
Board for Subscriptions: https://github.com/orgs/kartevonmorgen/projects/4?fullscreen=true
### Prerequisites
- [x] Use an external mail service kartevonmorgen/openfairdb#259
- [x] https://github.com/kartevonmorgen/openfairdb/issues/202
- [x] https://github.com/kartevonmorgen/openfairdb/issues/194
- [x] https://github.com/kartevonmorgen/openfairdb/issues/166 and 

### Target Group
Only a few power users will use the subscriptions, but those are the users, that improve the quality. So If we empower them, we improve the map qualitiy to all users, thats why it is a feature für all users.

### Problem
- Notifications are difficult to activate and if they come, there are too many with little intrest. To get more people beeing involved in mapping and ensuring content quality
- Its much too complicated to delete or to modify a subscription.

## Solution
- [x] Add Subscribe-Button [#443](https://github.com/kartevonmorgen/kartevonmorgen.ts/issues/151)

### Just allow to subscribe to any search and map view

- [ ] Rather usefull i think would be the option to **subscribe to hashtags**. So on the page where you choose the area, you also have the searchline to put in some keywords and tags. Especially if there are coming the events. But also like this it would help the theme-pilots to subscribe to alle csa-projects in the world. but only csa. https://github.com/kartevonmorgen/kvmflows/issues/5

- [ ] Subscribe to single entries, also **while editing**. https://github.com/kartevonmorgen/kartevonmorgen/issues/635 

- [ ] We should definitly include **Positive Impact Ratings** in the notification-System.  https://github.com/kartevonmorgen/kvmflows/issues/7 (More imprtant than the question, if an entry is on the map, is the question, how good it is rated. )

- [ ] Notify users at maximum once a day: **Only send notification Mail after there was no editing for at least 30 min.** Otherwise we spam users too much with notifications if we send it evrey minute and someone is just trying how the map works...  Maybe even just once a day https://github.com/kartevonmorgen/kvmflows/issues/9

- [ ] **Unsybscribe link in every notification-mail** https://github.com/kartevonmorgen/kvmflows/issues/4 . Each subscription is saved so you can delete every single subscription with an link in the mail. 

### Fazit 
- from David:
However I find one menu better: simpler, I don't need to search where I can find something but everything is in one place, the `Kartenausschnitt abonnieren`-thing doesn't really fit under `share`, I'm rather "sharing" it with myself, and we might have more features in the future that don't really fit in these two categories and if we put them there anyway it gets confusing for users.
- Helmut:
I think the Hamburger Menu contains everything, that leaves the map. The sharing menu has everything, that works on the map with popups. 
As subscription is much simpler in the future and works as a popup, it should be in the sharing button.

### LINKED ISSUE #194: Extend subscription and notification service
Currently all e-mails (signup, subscription, password reset) are dispatched to a local `sendmail` instance. This has various restrictions and drawbacks:

- Poor performance, i.e. one external process invocation per e-Mail
- The synchronous invocation doesn't scale
- Sending e-mails doesn't work within the Docker container and is disabled

The current implementation should be replaced by dispatching the sending of e-mails to a service provider.

Overview: https://konghq.com/blog/list-of-10-email-apis/

- [MailGun](https://www.mailgun.com/)
- [SendGrid](http://sendgrid.com/)
- [Mandrill](http://mandrill.com/)
- [mailjet](https://www.mailjet.com/email-api/)
- [MailChimp](https://mailchimp.com/)
- ...

The implementation should define an abstract interfaces that allow switching from one provider to another if required and doesn't introduce any dependencies on the actual API into the use cases.

Related issues:
- kartevonmorgen/openfairdb#166 Decouple post-processing from requests
- kartevonmorgen/openfairdb#172 Email-format issue
- kartevonmorgen/kvmflows#9 Send notifications as daily summary
- kartevonmorgen/kvmflows#4 Add link to unsubscribe in notification mail
- kartevonmorgen/openfairdb#95 Changes in email-texts
- kartevonmorgen/openfairdb#68 Support email templates

## ToDo's
- [x] Design overall architecture 
- [x] Create a components diagram
- [x] Implement it
- [x] https://github.com/kartevonmorgen/openfairdb/pull/259

### PR DESCRIPTION
**Pull request by flosse**
This PR is WIP. Feel free to comment!

### CODE REVIEW

File: ofdb-gateways/src/mailgun.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+use ofdb_entities::{event::*, place::*};
+
+// TODO: move into ofdb-core crate
+/// The kind of notification.
+#[derive(Debug, PartialEq)]
+pub enum Notification {
+    PlaceAdded(Place),
```
Comment (uklotzde) [inline ofdb-gateways/src/mailgun.rs]: Added -> Created

File: ofdb-gateways/src/mailgun.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    PlaceUpdated(Place),
```
Comment (uklotzde) [inline ofdb-gateways/src/mailgun.rs]: Do we need both the previous and updated version?

File: ofdb-gateways/src/mailgun.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+/// A email notification manager.
+pub struct Manager {
+    pub api_key: String,
+    pub domain: String,
+    pub from_email: String,
+    pub to_email: String,
+}
+
+impl NotificationManager for Manager {
+    fn notify(&self, n: Notification) {
```
Comment (uklotzde) [inline ofdb-gateways/src/mailgun.rs:34]: TODO: Send notifications from a separate thread and return immediately from the trait method.

File: ofdb-gateways/src/mailgun.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                    .post(&self.api_url())
+                    .form(&params)
+                    .basic_auth("api", Some(&self.api_key))
+                    .send();
+                match res {
+                    Ok(res) => {
+                        if res.status().is_success() {
+                            debug!("Mail provider response: {:#?}", res);
+                        } else {
+                            error!("Could not send email: response status: {:?}", res.status());
```
Comment (uklotzde) [inline ofdb-gateways/src/mailgun.rs:57]: We should distinguish between a technical failure (Err, see below) and an error response (here). Both log messages start with the same text.

**Comment by flosse** [comment]
close in favor of #263

**Pull request closed without merging**